README.md
# shape-editor

Javascript library based on Raphael.js for editing shapes

See demo at http://ome.github.io/shape-editor/

Development
===========

We use Grunt for various tools.

Install Node from https://nodejs.org, then:

::

    $ cd shape-editor
    $ npm install

Install Grunt CLI as described on http://gruntjs.com/using-the-cli.


To concatenate js files into a single shape-editor.js file that is used in the app:

::

    $ grunt concat


CHANGELOG.md
4.0.0 (May 2018)
----------------

This version includes:

 - removed scaling of line-widths when you zoom (PR #11)

3.1.1 (April 2018)
------------------

This version includes:

 - used openmicroscopy scoped package. See https://docs.npmjs.com/getting-started/scoped-packages

3.1.0 (March 2018)
------------------

This version includes:

 - added support for showing and editing polygons and polylines (PR #8)
 - allowed drawing of polygons or polylines

3.0.0 (June 2017)
-----------------

This version includes:

 - used x, y, radiusX and radiusY instead of cx, cy, rx and ry for ellipses (PR #6)

2.0.0 (June 2017)
-----------------

This version includes:

 - migrated shape-editor repository to ome
 - added bounding box support to shapes
 - added selectShapesById() and deleteById()
 - handled transform attribute on ellipse JSON

1.0.0 (Nov 2015)
----------------

Initial release


Gruntfile.js
/* jshint node: true */

module.exports = function(grunt) {

    var sources = [
        "src/js/shapes/line.js",
        "src/js/shapes/rect.js",
        "src/js/shapes/ellipse.js",
        "src/js/shapes/polygon.js",
        "src/js/shapeManager.js"
    ];

    grunt.initConfig({
        clean: {
            dist: [
                'dist/css',
                'dist/js/*.js'
            ]
        },
        jshint: {
            all: [
                "Gruntfile.js",
                "src/js/*.js"
            ],
            options: {
              jshintrc: '.jshintrc'
            }
        },
        concat: {
            options: {
                banner: "//! BSD License. www.openmicroscopy.org\n\n" +
                    "//!  DO NOT EDIT THIS FILE! - Edit under src/js/*.js\n"+
                    "//!  created by $ grunt concat\n\n",
                process: true,
                stripBanners: true,
            },
            dist: {
                src:  [ "<banner>" ].concat(sources),
                dest: "dist/js/shape-editor.js"
            }
        },
        watch: {
            files: [ "src/js/*.js",  "src/js/**/*.js"],
            tasks: ["jshint", "concat"]
        }
    });

    // Load tasks
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-concat');
    grunt.loadNpmTasks('grunt-contrib-watch');

    // Register tasks
    grunt.registerTask('default', [
        'clean'
    ]);

};

src/js/docs.js
/*
// Copyright (C) 2015 University of Dundee & Open Microscopy Environment.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following
// conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following
// disclaimer in the documentation // and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived 
// from this software without specific prior written permission.

// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, 
// BUT NOT LIMITED TO,
// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
// THE COPYRIGHT HOLDER OR CONTRIBUTORS
// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
// PROCUREMENT OF SUBSTITUTE // GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT // LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/* globals ShapeManager: false */
/* globals console: false */

$(function() {

    var WIDTH = 512,
        HEIGHT = 512;

    var options = {};

    var shapesClipboard = [];

    // We can choose to display shapes only - No editing
    // options = {'readOnly': true};

    var shapeManager = new ShapeManager("shapesCanvas",
                                        WIDTH, HEIGHT,
                                        options);

    var zoomPercent = 100;

    // set state depending on what we want to do,
    // for example to create Rectangle

    $("input[name='state']").click(function(){
        var state = $(this).val();
        shapeManager.setState(state);
    });

    $("input[name='strokeColor']").click(function(){
        var strokeColor = $(this).val();
        shapeManager.setStrokeColor(strokeColor);
    });

    $("select[name='strokeWidth']").change(function(){
        var strokeWidth = $(this).val();
        shapeManager.setStrokeWidth(strokeWidth);
    });

    var updateZoom = function updateZoom() {
        $("#zoomDisplay").text(zoomPercent + " %");
        shapeManager.setZoom(zoomPercent);
        var w = WIDTH * zoomPercent / 100,
            h = HEIGHT * zoomPercent / 100;
        $(".imageWrapper img").css({'width': w, 'height': h});
    };

    $("button[name='zoomIn']").click(function(){
        zoomPercent += 20;
        updateZoom();
    });
    $("button[name='zoomOut']").click(function(){
        zoomPercent -= 20;
        updateZoom();
    });

    $("button[name='deleteSelected']").click(function(){
        shapeManager.deleteSelectedShapes();
    });

    $("button[name='deleteAll']").click(function(){
        shapeManager.deleteAllShapes();
    });

    $("button[name='selectAll']").click(function(){
        shapeManager.selectAllShapes();
    });

    $("button[name='copyShapes']").click(function(){
        shapesClipboard = shapeManager.getSelectedShapesJson();
    });

    $("button[name='pasteShapes']").click(function(){
        // paste shapes, constraining to the image coords
        var p = shapeManager.pasteShapesJson(shapesClipboard, true);
        if (!p) {
            console.log("Shape could not be pasted: outside view port");
        }
    });

    $("button[name='getShapes']").click(function(){
      var json = shapeManager.getShapesJson();
      console.log(json);
    });

    $("button[name='getBBoxes']").click(function(){
      var shapes = shapeManager.getShapesJson();
      shapes.forEach(function(shape){
        var bbox = shapeManager.getShapeBoundingBox(shape.id);
        // Add each bbox as a Rectangle to image
        bbox.type = "Rectangle";
        bbox.strokeColor = "#ffffff";
        shapeManager.addShapeJson(bbox);
      });
    });

    $("button[name='selectShape']").click(function(){
      shapeManager.selectShapesById(1234);
    });

    var lastShapeId;
    $("button[name='deleteShapesByIds']").click(function(){
      shapeManager.deleteShapesByIds([lastShapeId]);
    });

    $("button[name='setShapes']").click(function(){
        var shapesJson = [
          {"type": "Rectangle",
            "strokeColor": "#ff00ff",
            "strokeWidth": 10,
            "x": 100, "y": 250,
            "width": 325, "height": 250},
          {"type": "Ellipse",
            "x": 300, "y": 250,
            "radiusX": 125, "radiusY": 250,
            "rotation": 100}
          ];
        shapeManager.setShapesJson(shapesJson);
    });

    $("#shapesCanvas").bind("change:selected", function(){
        var strokeColor = shapeManager.getStrokeColor();
        if (strokeColor) {
          $("input[value='" + strokeColor + "']").prop('checked', 'checked');
        } else {
           $("input[name='strokeColor']").removeProp('checked');
        }
        var strokeWidth = shapeManager.getStrokeWidth() || 1;
        $("select[name='strokeWidth']").val(strokeWidth);
    });

    $("#shapesCanvas").bind("change:shape", function(event, shapes){
        console.log("changed", shapes);
    });

    $("#shapesCanvas").bind("new:shape", function(event, shape){
        console.log("new", shape.toJson());
        console.log("selected", shapeManager.getSelectedShapesJson());
    });

    // Add some shapes to display
    shapeManager.addShapeJson({"type": "Polygon",
                               "points": "329,271 295,314 295,365 333,432 413,400 452,350 432,292 385,256",
                               "strokeColor": "#ffffff",
                               "strokeWidth": 0.5});

    shapeManager.addShapeJson({"type": "Polyline",
                               "points": "29,71 95,14 95,65 33,132 113,100 152,50",
                               "strokeColor": "#00ffdd",
                               "strokeWidth": 4});

    shapeManager.addShapeJson({"id": 1234,
                               "type": "Rectangle",
                               "strokeColor": "#ff00ff",
                               "strokeWidth": 6,
                               "x": 200, "y": 150,
                               "width": 125, "height": 150});

    shapeManager.addShapeJson({"type": "Ellipse",
                               "x": 200, "y": 150,
                               "radiusX": 125, "radiusY": 50,
                               "rotation": 45});

    shapeManager.addShapeJson({"type": "Ellipse",
                               "strokeColor": "#ffffff",
                               "x": 204, "y": 260,
                               "radiusX": 95, "radiusY": 55,
                               "transform": "matrix(0.82 0.56 -0.56 0.82 183.0 -69.7)"});

    shapeManager.addShapeJson({"type": "Arrow",
                               "strokeColor": "#ffff00",
                               "strokeWidth": 4,
                               "x1": 25, "y1": 450,
                               "x2": 200, "y2": 400});

    shapeManager.addShapeJson({"type": "Arrow",
                               "strokeColor": "#ffff00",
                               "strokeWidth": 10,
                               "x1": 25, "y1": 250,
                               "x2": 200, "y2": 200});

    shapeManager.addShapeJson({"type": "Ellipse",
                              "strokeColor": "#00ff00",
                              "radiusY": 31.5,
                              "radiusX": 91,
                              "transform": "matrix(2.39437435854 -0.644012141633 2.14261951162 0.765696311828 -1006.17788921 153.860479773)",
                              "strokeWidth": 2,
                              "y": 297.5,
                              "x": 258});

    shapeManager.addShapeJson({"type": "Ellipse",
                          "strokeColor": "#ffff00",
                          "radiusY": 71.5,
                          "radiusX": 41,
                          "transform": "matrix(0.839800601976 0.542894970432 -0.542894970432 0.839800601976 111.894472287 -140.195845758)",
                          "strokeWidth": 2,
                          "y": 260.5,
                          "x": 419});

    var s = shapeManager.addShapeJson({"type": "Line",
                               "strokeColor": "#00ff00",
                               "strokeWidth": 2,
                               "x1": 400, "y1": 400,
                               "x2": 250, "y2": 310});
    lastShapeId = s.toJson().id;

    // We start off in the 'SELECT' mode
    shapeManager.setState("SELECT");
});


src/js/shapeEditor.js

/*
// Copyright (C) 2015 University of Dundee & Open Microscopy Environment.
// All rights reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/* globals Raphael: false */
/* globals console: false */

var ShapeEditor = function ShapeEditor(elementId, width, height, options) {

	this.STATES = ["SELECT", "RECT", "LINE", "ARROW", "ELLIPSE"];
    // Set up Raphael paper...
    this.paper = Raphael(elementId, width, height);

    // Add a full-size background to cover existing shapes while
    // we're creating new shapes, to stop them being selected.
    // Mouse events on this will bubble up to svg and are handled below
    this.newShapeBg = this.paper.rect(0, 0, width, height);
    this.newShapeBg.attr({'fill':'#000',
                          'fill-opacity':0.01,
                          'cursor': 'crosshair'});
};


ShapeEditor.prototype.setState = function setState(state) {
	if (this.STATES.indexOf(state) === -1) {
        console.log("Invalid state: ", state, "Needs to be in", this.STATES);
        return;
    }
    // When creating shapes, cover existing shapes with newShapeBg
    var shapes = ["RECT", "LINE", "ARROW", "ELLIPSE"];
    if (shapes.indexOf(state) > -1) {
        this.newShapeBg.show().toFront();
    } else {
        this.newShapeBg.hide();
    }
    this.state = state;
};

ShapeEditor.prototype.getState = function getState() {
    return this.state;
};


src/js/shapes/ellipse.js
/*
// Copyright (C) 2015 University of Dundee & Open Microscopy Environment.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following
// conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following
// disclaimer in the documentation // and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived 
// from this software without specific prior written permission.

// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, 
// BUT NOT LIMITED TO,
// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
// THE COPYRIGHT HOLDER OR CONTRIBUTORS
// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
// PROCUREMENT OF SUBSTITUTE // GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT // LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/* globals Raphael: false */
/* globals console: false */

var Ellipse = function Ellipse(options) {

    var self = this;
    this.manager = options.manager;
    this.paper = options.paper;

    if (options.id) {
        this._id = options.id;
    } else {
        this._id = this.manager.getRandomId();
    }
    this._x = options.x;
    this._y = options.y;
    this._radiusX = options.radiusX;
    this._radiusY = options.radiusY;
    this._rotation = options.rotation || 0;

    // We handle transform matrix by creating this.Matrix
    // This is used as a one-off transform of the handles positions
    // when they are created. This then updates the _x, _y, _radiusX, _radiusY & rotation
    // of the ellipse itself (see below)
    if (options.transform && options.transform.startsWith('matrix')) {
        var tt = options.transform.replace('matrix(', '').replace(')', '').split(" ");
        var a1 = parseFloat(tt[0]);
        var a2 = parseFloat(tt[1]);
        var b1 = parseFloat(tt[2]);
        var b2 = parseFloat(tt[3]);
        var c1 = parseFloat(tt[4]);
        var c2 = parseFloat(tt[5]);
        this.Matrix = Raphael.matrix(a1, a2, b1, b2, c1, c2);
    }

    if (this._radiusX === 0 || this._radiusY === 0) {
        this._yxRatio = 0.5;
    } else {
        this._yxRatio = this._radiusY / this._radiusX;
    }

    this._strokeColor = options.strokeColor;
    this._strokeWidth = options.strokeWidth || 2;
    this._selected = false;
    this._zoomFraction = 1;
    if (options.zoom) {
        this._zoomFraction = options.zoom / 100;
    }
    this.handle_wh = 6;

    this.element = this.paper.ellipse();
    this.element.attr({'fill-opacity': 0.01,
                        'fill': '#fff',
                        'cursor': 'pointer'});

    // Drag handling of ellipse
    if (this.manager.canEdit) {
        this.element.drag(
            function(dx, dy) {
                // DRAG, update location and redraw
                dx = dx / self._zoomFraction;
                dy = dy / self._zoomFraction;

                var offsetX = dx - this.prevX;
                var offsetY = dy - this.prevY;
                this.prevX = dx;
                this.prevY = dy;

                // Manager handles move and redraw
                self.manager.moveSelectedShapes(offsetX, offsetY, true);
                return false;
            },
            function() {
                // START drag: note the start location
                self._handleMousedown();
                this.prevX = 0;
                this.prevY = 0;
                return false;
            },
            function() {
                // STOP
                // notify changed if moved
                if (this.prevX !== 0 || this.prevY !== 0) {
                    self.manager.notifySelectedShapesChanged();
                }
                return false;
            }
        );
    }

    // create handles, applying this.Matrix if set
    this.createHandles();
    // update x, y, radiusX, radiusY & rotation
    // If we have Matrix, recalculate width/height ratio based on all handles
    var resizeWidth = !!this.Matrix;
    this.updateShapeFromHandles(resizeWidth);
    // and draw the Ellipse
    this.drawShape();
};

Ellipse.prototype.toJson = function toJson() {
    var rv = {
        'type': "Ellipse",
        'x': this._x,
        'y': this._y,
        'radiusX': this._radiusX,
        'radiusY': this._radiusY,
        'rotation': this._rotation,
        'strokeWidth': this._strokeWidth,
        'strokeColor': this._strokeColor
    };
    if (this._id) {
        rv.id = this._id;
    }
    return rv;
};

Ellipse.prototype.compareCoords = function compareCoords(json) {

    var selfJson = this.toJson(),
        match = true;
    if (json.type !== selfJson.type) {
        return false;
    }
    ['x', 'y', 'radiusX', 'radiusY', 'rotation'].forEach(function(c){
        if (Math.round(json[c]) !== Math.round(selfJson[c])) {
            match = false;
        }
    });
    return match;
};

// Useful for pasting json with an offset
Ellipse.prototype.offsetCoords = function offsetCoords(json, dx, dy) {
    json.x = json.x + dx;
    json.y = json.y + dy;
    return json;
};

// Shift this shape by dx and dy
Ellipse.prototype.offsetShape = function offsetShape(dx, dy) {
    this._x = this._x + dx;
    this._y = this._y + dy;
    this.drawShape();
};

// handle start of drag by selecting this shape
// if not already selected
Ellipse.prototype._handleMousedown = function _handleMousedown() {
    if (!this._selected) {
        this.manager.selectShapes([this]);
    }
};

Ellipse.prototype.setColor = function setColor(strokeColor) {
    this._strokeColor = strokeColor;
    this.drawShape();
};

Ellipse.prototype.getStrokeColor = function getStrokeColor() {
    return this._strokeColor;
};

Ellipse.prototype.setStrokeColor = function setStrokeColor(strokeColor) {
    this._strokeColor = strokeColor;
    this.drawShape();
};

Ellipse.prototype.setStrokeWidth = function setStrokeWidth(strokeWidth) {
    this._strokeWidth = strokeWidth;
    this.drawShape();
};

Ellipse.prototype.getStrokeWidth = function getStrokeWidth() {
    return this._strokeWidth;
};

Ellipse.prototype.destroy = function destroy() {
    this.element.remove();
    this.handles.remove();
};

Ellipse.prototype.intersectRegion = function intersectRegion(region) {
    var path = this.manager.regionToPath(region, this._zoomFraction * 100);
    var f = this._zoomFraction,
        x = parseInt(this._x * f, 10),
        y = parseInt(this._y * f, 10);

    if (Raphael.isPointInsidePath(path, x, y)) {
        return true;
    }
    var path2 = this.getPath(),
        i = Raphael.pathIntersection(path, path2);
    return (i.length > 0);
};

Ellipse.prototype.getPath = function getPath() {

    // Adapted from https://github.com/poilu/raphael-boolean
    var a = this.element.attrs,
        radiusX = a.radiusX,
        radiusY = a.radiusY,
        cornerPoints = [
            [a.x - radiusX, a.y - radiusY],
            [a.x + radiusX, a.y - radiusY],
            [a.x + radiusX, a.y + radiusY],
            [a.x - radiusX, a.y + radiusY]
        ],
        path = [];
    var radiusShift = [
        [
            [0, 1],
            [1, 0]
        ],
        [
            [-1, 0],
            [0, 1]
        ],
        [
            [0, -1],
            [-1, 0]
        ],
        [
            [1, 0],
            [0, -1]
        ]
    ];

    //iterate all corners
    for (var i = 0; i <= 3; i++) {
        //insert starting point
        if (i === 0) {
            path.push(["M", cornerPoints[0][0], cornerPoints[0][1] + radiusY]);
        }

        //insert "curveto" (radius factor .446 is taken from Inkscape)
        var c1 = [cornerPoints[i][0] + radiusShift[i][0][0] * radiusX * 0.446, cornerPoints[i][1] + radiusShift[i][0][1] * radiusY * 0.446];
        var c2 = [cornerPoints[i][0] + radiusShift[i][1][0] * radiusX * 0.446, cornerPoints[i][1] + radiusShift[i][1][1] * radiusY * 0.446];
        var p2 = [cornerPoints[i][0] + radiusShift[i][1][0] * radiusX, cornerPoints[i][1] + radiusShift[i][1][1] * radiusY];
        path.push(["C", c1[0], c1[1], c2[0], c2[1], p2[0], p2[1]]);
    }
    path.push(["Z"]);
    path = path.join(",").replace(/,?([achlmqrstvxz]),?/gi, "$1");

    if (this._rotation !== 0) {
        path = Raphael.transformPath(path, "r" + this._rotation);
    }
    return path;
};

Ellipse.prototype.isSelected = function isSelected() {
    return this._selected;
};

Ellipse.prototype.setZoom = function setZoom(zoom) {
    this._zoomFraction = zoom / 100;
    this.drawShape();
};

Ellipse.prototype.updateHandle = function updateHandle(handleId, x, y, shiftKey) {
    // Refresh the handle coordinates, then update the specified handle
    // using MODEL coordinates
    this._handleIds = this.getHandleCoords();
    var h = this._handleIds[handleId];
    h.x = x;
    h.y = y;
    var resizeWidth = (handleId === "left" || handleId === "right");
    this.updateShapeFromHandles(resizeWidth, shiftKey);
};

Ellipse.prototype.updateShapeFromHandles = function updateShapeFromHandles(resizeWidth, shiftKey) {
    var hh = this._handleIds,
        lengthX = hh.end.x - hh.start.x,
        lengthY = hh.end.y - hh.start.y,
        widthX = hh.left.x - hh.right.x,
        widthY = hh.left.y - hh.right.y,
        rot;
    // Use the 'start' and 'end' handles to get rotation and length
    if (lengthX === 0){
        this._rotation = 90;
    } else if (lengthX > 0) {
        rot = Math.atan(lengthY / lengthX);
        this._rotation = Raphael.deg(rot);
    } else if (lengthX < 0) {
        rot = Math.atan(lengthY / lengthX);
        this._rotation = 180 + Raphael.deg(rot);
    }
    
    // centre is half-way between 'start' and 'end' handles
    this._x = (hh.start.x + hh.end.x)/2;
    this._y = (hh.start.y + hh.end.y)/2;
    // Radius-x is half of distance between handles
    this._radiusX = Math.sqrt((lengthX * lengthX) + (lengthY * lengthY)) / 2;
    // Radius-y may depend on handles OR on x/y ratio
    if (resizeWidth) {
        this._radiusY = Math.sqrt((widthX * widthX) + (widthY * widthY)) / 2;
        this._yxRatio = this._radiusY / this._radiusX;
    } else {
        if (shiftKey) {
            this._yxRatio = 1;
        }
        this._radiusY = this._yxRatio * this._radiusX;
    }

    this.drawShape();
};

Ellipse.prototype.drawShape = function drawShape() {

    var strokeColor = this._strokeColor,
        strokeW = this._strokeWidth;

    var f = this._zoomFraction,
        x = this._x * f,
        y = this._y * f,
        radiusX = this._radiusX * f,
        radiusY = this._radiusY * f;

    this.element.attr({'cx': x,
                       'cy': y,
                       'rx': radiusX,
                       'ry': radiusY,
                       'stroke': strokeColor,
                       'stroke-width': strokeW});
    this.element.transform('r'+ this._rotation);

    if (this.isSelected()) {
        this.element.toFront();
        this.handles.show().toFront();
    } else {
        this.handles.hide();
    }

    // handles have been updated (model coords)
    this._handleIds = this.getHandleCoords();
    var hnd, h_id, hx, hy;
    for (var h=0, l=this.handles.length; h<l; h++) {
        hnd = this.handles[h];
        h_id = hnd.h_id;
        hx = this._handleIds[h_id].x * this._zoomFraction;
        hy = this._handleIds[h_id].y * this._zoomFraction;
        hnd.attr({'x':hx-this.handle_wh/2, 'y':hy-this.handle_wh/2});
    }
};

Ellipse.prototype.setSelected = function setSelected(selected) {
    this._selected = !!selected;
    this.drawShape();
};


Ellipse.prototype.createHandles = function createHandles() {
    // ---- Create Handles -----

    // NB: handleIds are used to calculate ellipse coords
    // so handledIds are scaled to MODEL coords, not zoomed.
    this._handleIds = this.getHandleCoords();

    var self = this,
        // map of centre-points for each handle
        handleAttrs = {'stroke': '#4b80f9',
                        'fill': '#fff',
                        'cursor': 'move',
                        'fill-opacity': 1.0};

    // draw handles
    self.handles = this.paper.set();
    var _handle_drag = function() {
        return function (dx, dy, mouseX, mouseY, event) {
            dx = dx / self._zoomFraction;
            dy = dy / self._zoomFraction;
            // on DRAG...
            var absX = dx + this.ox,
                absY = dy + this.oy;
            self.updateHandle(this.h_id, absX, absY, event.shiftKey);
            return false;
        };
    };
    var _handle_drag_start = function() {
        return function () {
            // START drag: simply note the location we started
            // we scale by zoom to get the 'model' coordinates
            this.ox = (this.attr("x") + this.attr('width')/2) / self._zoomFraction;
            this.oy = (this.attr("y") + this.attr('height')/2) / self._zoomFraction;
            return false;
        };
    };
    var _handle_drag_end = function() {
        return function() {
            // simply notify manager that shape has changed
            self.manager.notifyShapesChanged([self]);
            return false;
        };
    };

    var hsize = this.handle_wh,
        hx, hy, handle;
    for (var key in this._handleIds) {
        hx = this._handleIds[key].x;
        hy = this._handleIds[key].y;
        // If we have a transformation matrix, apply it...
        if (this.Matrix) {
            var matrixStr = this.Matrix.toTransformString();
            // Matrix that only contains rotation and translation 
            // E.g. t111.894472287,-140.195845758r32.881,0,0  Will be handled correctly:
            // Resulting handles position and x,y radii will be calculated
            // so we don't need to apply transform to ellipse itself
            // BUT, if we have other transforms such as skew, we can't do this.
            // Best to just show warning if Raphael can't resolve matrix to simpler transforms:
            // E.g. m2.39,-0.6,2.1,0.7,-1006,153 
            if (matrixStr.indexOf('m') > -1) {
                console.log("Matrix only supports rotation & translation. " + matrixStr + " may contain skew for shape: ", this.toJson());
            }
            var mx = this.Matrix.x(hx, hy);
            var my = this.Matrix.y(hx, hy);
            hx = mx;
            hy = my;
            // update the source coordinates
            this._handleIds[key].x = hx;
            this._handleIds[key].y = hy;
        }
        handle = this.paper.rect(hx-hsize/2, hy-hsize/2, hsize, hsize);
        handle.attr({'cursor': 'move'});
        handle.h_id = key;
        handle.line = self;

        if (this.manager.canEdit) {
            handle.drag(
                _handle_drag(),
                _handle_drag_start(),
                _handle_drag_end()
            );
        }
        self.handles.push(handle);
    }

    self.handles.attr(handleAttrs).hide();     // show on selection
};

Ellipse.prototype.getHandleCoords = function getHandleCoords() {
    // Returns MODEL coordinates (not zoom coordinates)
    var rot = Raphael.rad(this._rotation),
        x = this._x,
        y = this._y,
        radiusX = this._radiusX,
        radiusY = this._radiusY,
        startX = x - (Math.cos(rot) * radiusX),
        startY = y - (Math.sin(rot) * radiusX),
        endX = x + (Math.cos(rot) * radiusX),
        endY = y + (Math.sin(rot) * radiusX),
        leftX = x + (Math.sin(rot) * radiusY),
        leftY = y - (Math.cos(rot) * radiusY),
        rightX = x - (Math.sin(rot) * radiusY),
        rightY = y + (Math.cos(rot) * radiusY);

    return {'start':{x: startX, y: startY},
            'end':{x: endX, y: endY},
            'left':{x: leftX, y: leftY},
            'right':{x: rightX, y: rightY}
        };
};


// Class for creating Lines.
var CreateEllipse = function CreateEllipse(options) {

    this.paper = options.paper;
    this.manager = options.manager;
};

CreateEllipse.prototype.startDrag = function startDrag(startX, startY) {

    var strokeColor = this.manager.getStrokeColor(),
        strokeWidth = this.manager.getStrokeWidth(),
        zoom = this.manager.getZoom();

    this.ellipse = new Ellipse({
        'manager': this.manager,
        'paper': this.paper,
        'x': startX,
        'y': startY,
        'radiusX': 0,
        'radiusY': 0,
        'rotation': 0,
        'strokeWidth': strokeWidth,
        'zoom': zoom,
        'strokeColor': strokeColor});
};

CreateEllipse.prototype.drag = function drag(dragX, dragY, shiftKey) {

    this.ellipse.updateHandle('end', dragX, dragY, shiftKey);
};

CreateEllipse.prototype.stopDrag = function stopDrag() {

    // Don't create ellipse of zero size (click, without drag)
    var coords = this.ellipse.toJson();
    if (coords.radiusX < 2) {
        this.ellipse.destroy();
        delete this.ellipse;
        return;
    }
    // on the 'new:shape' trigger, this shape will already be selected
    this.ellipse.setSelected(true);
    this.manager.addShape(this.ellipse);
};
